Remove validation of SSL_CTX_set_options

andreas-bok-sociomantic opened the pull request:
SSL_CTX_set_options returns the new/updated bitmask defining the current options which are set. 

Hence, we can't compare the return value against 1 to check whether the options were set correctly.

andreas-bok-sociomantic committed:
Fix validation of SSL_CTX_set_options

SSL_CTX_set_options returns the updated bitmask
defining the options which are currently set.
Hence, the bitwise operator & should be used to
check whether the options were set correctly.

federico-cuello-sociomantic reviewed (approved):
Yeah, sorry. I have that in my own fork.

ben-palmer-sociomantic commented on src/ssl.cc:
```diff
@@ -47,11 +47,7 @@
 drizzle_return_t drizzle_set_ssl(drizzle_st *con, const char *key, const char *cert, const char *ca, const char *capath, const char *cipher)
```
Could you rephrase the commit message, I am struggling to work out what you mean.

andreas-bok-sociomantic replied:
Yes

andreas-bok-sociomantic commented:
Updated ci msg

ben-palmer-sociomantic commented:
Cool, commit message is clear now. One more question, should we be doing another check on the returned bitmask now?

andreas-bok-sociomantic commented:
> Cool, commit message is clear now. One more question, should we be doing another check on the returned bitmask now?

Yes and no. 
- Definitely safer to check that it actually is set correctly. I will add a check. But
- Though not currently supported there are several other SSL options which can be set. Thus, we would need to compare it against the user specified options.

andreas-bok-sociomantic commented:
- Fixed the code to check if ssl options are correctly set
- Updated the ci msg 

Previous head was 5ef01a3, new **HEAD** at 9d7a0c2.

<details>
<summary>Diff between heads are:</summary>

```diff

 src/ssl.cc | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/src/ssl.cc b/src/ssl.cc
index e3fcd95..fe088c5 100644
--- a/src/ssl.cc
+++ b/src/ssl.cc
@@ -47,7 +47,14 @@
 drizzle_return_t drizzle_set_ssl(drizzle_st *con, const char *key, const char *cert, const char *ca, const char *capath, const char *cipher)
 {
   con->ssl_context= SSL_CTX_new(SSLv23_client_method());
-  SSL_CTX_set_options((SSL_CTX*)con->ssl_context, SSL_OP_NO_SSLv2 | SSL_OP_NO_SSLv3);
+
+  long ssl_options = SSL_CTX_set_options((SSL_CTX *)con->ssl_context,
+                                         SSL_OP_NO_SSLv2 | SSL_OP_NO_SSLv3);
+  if (!(ssl_options & (SSL_OP_NO_SSLv2 | SSL_OP_NO_SSLv3)))
+  {
+    drizzle_set_error(con, __FILE_LINE_FUNC__, "Cannot set the SSL protocol options");
+    return DRIZZLE_RETURN_SSL_ERROR;
+  }
 
   if (cipher)
   {

```
</details>

ben-palmer-sociomantic commented on src/ssl.cc:
```diff
@@ -47,7 +47,10 @@
+
+  long ssl_options = SSL_CTX_set_options((SSL_CTX *)con->ssl_context,
+                                         SSL_OP_NO_SSLv2 | SSL_OP_NO_SSLv3);
+  if (!(ssl_options & (SSL_OP_NO_SSLv2 | SSL_OP_NO_SSLv3)))
```
Would this be simpler as:
```
long required_ssl_options = (SSL_OP_NO_SSLv2 | SSL_OP_NO_SSLv3);
long set_ssl_options = SSL_CTX_set_options((SSL_CTX *)con->ssl_context, required_ssl_options);
if (set_ssl_options != required_ssl_options)
```

andreas-bok-sociomantic replied:
> if (set_ssl_options != required_ssl_options)

We still need to do bitwise & since the returned options bitmask also contains the default options.

ben-palmer-sociomantic reviewed (approved):
One optional comment otherwise LGTM.

andreas-bok-sociomantic commented:
Updated based on comments from @ben-palmer-sociomantic 

Previous head was 9d7a0c2, new **HEAD** at 466dc85.

<details>
<summary>Diff between heads are:</summary>

```diff

 src/ssl.cc | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/src/ssl.cc b/src/ssl.cc
index fe088c5..8aa75be 100644
--- a/src/ssl.cc
+++ b/src/ssl.cc
@@ -48,9 +48,10 @@ drizzle_return_t drizzle_set_ssl(drizzle_st *con, const char *key, const char *c
 {
   con->ssl_context= SSL_CTX_new(SSLv23_client_method());
 
-  long ssl_options = SSL_CTX_set_options((SSL_CTX *)con->ssl_context,
-                                         SSL_OP_NO_SSLv2 | SSL_OP_NO_SSLv3);
-  if (!(ssl_options & (SSL_OP_NO_SSLv2 | SSL_OP_NO_SSLv3)))
+  const long required_ssl_options = (SSL_OP_NO_SSLv2 | SSL_OP_NO_SSLv3);
+  long ssl_options =
+      SSL_CTX_set_options((SSL_CTX *)con->ssl_context, required_ssl_options);
+  if (!(ssl_options & required_ssl_options))
   {
     drizzle_set_error(con, __FILE_LINE_FUNC__, "Cannot set the SSL protocol options");
     return DRIZZLE_RETURN_SSL_ERROR;

```
</details>

ben-palmer-sociomantic commented:
LGTM
